File: frontend/src/domain/Contract.ts
export type ContractPaymentStatus = "unpaid" | "invoiced" | "partial" | "paid";

export interface ContractStage {
  id: number;
  sortOrder: number;
  description: string;
  amount: number;
  status: "pending" | "invoiced" | "paid";
  paidAt?: string | null;
  /** "subscription" activates recurring billing when paid; "one_time" is a plain fee. */
  kind?: "subscription" | "one_time";
}

export interface Contract {
  id: number;
  accountId: number;
  title: string;
  body: string;
  totalAmount: number;
  currency: string;
  status: "draft" | "pending_signature" | "signed" | "voided";
  signToken: string;
  signedAt: string | null;
  signerName: string | null;
  signerPosition?: string | null;
  signedCopyEmail?: string | null;
  signaturePngBase64: string | null;
  pdfBase64: string | null;
  createdAt: string;
  stages: ContractStage[];
  billingInstructionId?: number | null;
  monthlyAmount?: number | null;
  subscriptionMonths?: number | null;
  billingDay?: number | null;
  /** Derived from the stages — never stored, so it cannot drift. */
  paymentStatus?: ContractPaymentStatus;
  /** Whole-contract worth: monthly x months + fees. Null when open-ended. */
  contractValue?: number | null;
}

export interface ContractPublic {
  id: number;
  title: string;
  body: string;
  totalAmount: number;
  currency: string;
  status: string;
  signedAt: string | null;
  signerName: string | null;
  pdfBase64: string | null;
  stages: ContractStage[];
  monthlyAmount?: number | null;
  subscriptionStatus?: string | null;
  paymentStatus?: ContractPaymentStatus;
  contractValue?: number | null;
}

export interface ContractStageInput {
  description: string;
  amount: number;
}

export interface ContractCreateInput {
  accountId: number;
  title: string;
  body: string;
  currency: string;
  pdfBase64?: string | null;
  stages: ContractStageInput[];
  isSubscription?: boolean;
  monthlyAmount?: number | null;
  subscriptionMonths?: number | null;
  billingDay?: number | null;
}

export interface ContractMemberRow {
  id: number;
  title: string;
  totalAmount: number;
  currency: string;
  status: Contract["status"];
  signToken: string;
  signedAt: string | null;
  signerName: string | null;
  createdAt: string;
  stages: ContractStage[];
  monthlyAmount?: number | null;
  subscriptionMonths?: number | null;
}

File: backend/app/models/contracts/contract.py
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.session import Base


class Contract(Base):
    """Admin-created contract for an account with payment stages and signature."""

    __tablename__ = "contracts"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    account_id: Mapped[int] = mapped_column(
        ForeignKey("accounts.id", ondelete="CASCADE"), nullable=False, index=True
    )

    title: Mapped[str] = mapped_column(String(500), nullable=False)
    body: Mapped[str] = mapped_column(Text, nullable=False, default="")

    total_amount: Mapped[float] = mapped_column(Float, nullable=False)
    currency: Mapped[str] = mapped_column(String(8), nullable=False, default="ILS")

    # "draft" | "pending_signature" | "signed" | "voided"
    status: Mapped[str] = mapped_column(String(32), nullable=False, default="draft")

    # Link to billing instruction for monthly subscriptions
    billing_instruction_id: Mapped[int | None] = mapped_column(
        ForeignKey("account_billing_instructions.id", ondelete="SET NULL"), nullable=True
    )
    # Monthly subscription amount (if this is a subscription contract)
    monthly_amount: Mapped[float | None] = mapped_column(Float, nullable=True)
    subscription_months: Mapped[int | None] = mapped_column(Integer, nullable=True)

    # Token used in the public signing URL — never changes once created
    sign_token: Mapped[str] = mapped_column(
        String(64), nullable=False, unique=True, default=lambda: uuid.uuid4().hex
    )

    # Optional PDF attachment shown to the client during signing
    pdf_base64: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Populated when client signs
    signature_png_base64: Mapped[str | None] = mapped_column(Text, nullable=True)
    signed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    signer_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    signer_position: Mapped[str | None] = mapped_column(String(255), nullable=True)
    signed_copy_email: Mapped[str | None] = mapped_column(String(255), nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    created_by_admin_id: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )

    stages: Mapped[list[ContractPaymentStage]] = relationship(
        "ContractPaymentStage",
        back_populates="contract",
        order_by="ContractPaymentStage.sort_order",
        cascade="all, delete-orphan",
    )


from .contract_payment_stage import ContractPaymentStage  # noqa: E402
